### ISSUE TITLE
Equivalent of caffe iter_size in TF

### ISSUE THREAD

**Issue by paragon00** [tensorflow/tensorflow#14165]
Can we get the equivalent of caffe's iter_size parameter in TF? This accumulates gradient calcs over several GPU cycles before doing the weight update. It effectively allows a larger batch size. TensorFlow doesn't natively have this but some ppl seem to have implemented sth like it themselves, e.g

https://stackoverflow.com/questions/42156957/how-to-update-model-parameters-with-accumulated-gradients

I think it'd be a useful parameter to have as part of official TF ... or is there some easy way to implement this functionality already?

Shaun

**Comment by bignamehyp** [comment]
You can achieve this using tf control flows. .I'm going to mark this as "contributions welcome" since I think it would be a nice feature and would benefit the whole community.

**Comment by REFunction** [comment]
I really want a iter_size, or I have to turn to pytorch
Anyone could do something?

**Comment by ppwwyyxx** [comment]
> Anyone could do something?

It's easy to implement and you can take the code from here:
https://github.com/tensorpack/tensorpack/blob/45ebac959f34507f29176fc12d327f3cc9ff7468/tensorpack/tfutils/optimizer.py#L133-L202

**Comment by REFunction** [comment]
> > Anyone could do something?
> 
> It's easy to implement and you can take the code from here:
> https://github.com/tensorpack/tensorpack/blob/45ebac959f34507f29176fc12d327f3cc9ff7468/tensorpack/tfutils/optimizer.py#L133-L202

Thank you for providing so excellent repo. Now I have installed and I am preparing for testing on my previous code. If there is no problem, I'd like to introduce it to others.

**Comment by ppwwyyxx** [comment]
Thanks. To use this optimizer it takes is only one extra line of code:
```python
        from tensorpack.tfutils.optimizer import AccumGradOptimizer
        myopt = tf.train.GradientDescentOptimizer(0.01)
        myopt = AccumGradOptimizer(myopt, niter=5)
        train_op = myopt.minimize(loss)
```

**Comment by REFunction** [comment]
> Thanks. To use this optimizer it takes only one extra line of code:
> 
> ```python
>         from tensorpack.tfutils.optimizer import AccumGradOptimizer
>         myopt = tf.train.GradientDescentOptimizer(0.01)
>         myopt = AccumGradOptimizer(myopt, niter=5)
>         train_op = myopt.minimize(loss)
> ```

OK,  got it and I am running a program with AccumGradOptimizer. I worry that batch norm will go wrong because someone before gave his code with the same target but I came across many bugs. 
So I am testing.
Another problem is learning rate. I should enlarge learning rate according to niter, or not?

**Comment by ppwwyyxx** [comment]
> I worry that batch norm will go wrong 

If your per-iteration batch size is too small and make BatchNorm fail, `iter_size` will not help you.

>  I should enlarge learning rate according to niter, or not?

What my optimizer does is written in the code:
>     An optimizer which accumulates gradients across :math:`k` :meth:`minimize` calls,
    and apply them together in every :math:`k`th :meth:`minimize` call.
    This is roughly the same as using a :math:`k` times larger batch size plus a
    :math:`k` times larger learning rate, but uses much less memory.

Whether you want to adjust learning rate on top of that is up to you. Based on common folk wisdom you do not need to adjust learning rate further.

**Comment by REFunction** [comment]
> > I worry that batch norm will go wrong
> 
> If your per-iteration batch size is too small and make BatchNorm fail, `iter_size` will not help you.
> In that case you'll need cross-GPU BatchNorm (which is also supported in tensorpack), or GroupNorm as a replacement (which I authored).
> 
> > I should enlarge learning rate according to niter, or not?
> 
> What my optimizer does is written in the code:
> 
> > An optimizer which accumulates gradients across :math:`k` :meth:`minimize` calls,
> > and apply them together in every :math:`k`th :meth:`minimize` call.
> > This is roughly the same as using a :math:`k` times larger batch size plus a
> > :math:`k` times larger learning rate, but uses much less memory.
> 
> Whether you want to adjust learning rate on top of that is up to you. Based on common folk wisdom you do not need to adjust learning rate further.

Your repo helps me a lot. But I find that **Global Step** is not supported while some repos train networks with dynamic learning rate, or other tricks, where Global Step is necessary. So I would like to know whether it's difficult to implement with Global Step? Or other reasons?

**Comment by ppwwyyxx** [comment]
> But I find that Global Step is not supported while some repos train networks with dynamic learning rate, or other tricks, where Global Step is necessary. So I would like to know whether it's difficult to implement with Global Step? Or other reasons?

It takes about [three lines of code](https://github.com/tensorpack/tensorpack/commit/d8dd7ca93640c7299d2c2da064d57982fe56ff69) to support it.
It was not there earlier because tensorpack already manages global_step by other methods, rather than by the optimizer.

**Comment by REFunction** [comment]
> > But I find that Global Step is not supported while some repos train networks with dynamic learning rate, or other tricks, where Global Step is necessary. So I would like to know whether it's difficult to implement with Global Step? Or other reasons?
> 
> It takes about [three lines of code](https://github.com/tensorpack/tensorpack/commit/d8dd7ca93640c7299d2c2da064d57982fe56ff69) to support it.
> It was not there earlier because tensorpack already manages global_step by other methods, rather than by the optimizer.

WOW! So fast updating! What an enthusiastic man!Thank you.

**Comment by tilakrayal** [comment]
@paragon00,
Could you please take a look at this [comment](https://github.com/tensorflow/tensorflow/issues/14165#issuecomment-478323338) from the developer and try to update to the latest stable v2.9 and let us know if the issue still persists. Thank you!
